### PR TITLE
Martin/time machine layout transition issues

### PR DESCRIPTION
**Pull request by MartinJibber**
Significant improvements to the animation when switching between read and write modes.

https://user-images.githubusercontent.com/88675954/152101821-92d982df-d5b1-4925-90fe-920e517c2aa5.mp4

### CODE REVIEW

File: Shared/Flows/Onboarding/Welcome/WelcomeCollectionView.swift [hunk 1]
```diff
@@ -18,7 +18,7 @@ class WelcomeCollectionView: CollectionView {
     init() {
         let layout = MessagesTimeMachineCollectionViewLayout()
         layout.messageContentState = .collapsed
-        layout.uiState = .write
```
Comment (MartinJibber) [inline Shared/Flows/Onboarding/Welcome/WelcomeCollectionView.swift:21]: I cleaned up the layout object so it doesn't know anything about the different UI states. More on this later...

File: iOS/Flows/Conversation/Cells/ConversationMessagesCell.swift [hunk 2]
```diff
@@ -65,6 +65,7 @@ class ConversationMessagesCell: UICollectionViewCell, ConversationUIStateSettabl
     override init(frame: CGRect) {
         super.init(frame: frame)
 
+        self.configureCollectionLayout(for: .read)
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/ConversationMessagesCell.swift:68]: Now it's the responsibility of the class using the layout to sets the appropriate spacing and positioning parameters.

File: iOS/Flows/Conversation/Cells/ConversationMessagesCell.swift [hunk 3]
```diff
@@ -111,15 +112,28 @@ class ConversationMessagesCell: UICollectionViewCell, ConversationUIStateSettabl
...
-                self.scrollToLastItemOnLayout = true
-                self.setNeedsLayout()
-                self.collectionLayout.finalizeLayoutTransition()
-            })
+            await self.dataSource.reconfigureAllItems()
+            if state == .write {
+                let maxOffset = self.collectionLayout.maxZPosition
+                self.collectionView.setContentOffset(CGPoint(x: 0, y: maxOffset), animated: true)
+            }
+        }
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/ConversationMessagesCell.swift:123]: Updating the layout parameters then reconfiguring the items makes the animations look smooth. Attempting to animation a with layoutNow/setNeedsLayout is less successful.

File: iOS/Flows/Conversation/Cells/ConversationMessagesCell.swift [hunk 4]
```diff
@@ -111,15 +112,28 @@ class ConversationMessagesCell: UICollectionViewCell, ConversationUIStateSettabl
...
+        self.collectionLayout.itemHeight
+        = MessageContentView.bubbleHeight + MessageDetailView.height + Theme.ContentOffset.short.value
+
+        switch state {
+        case .read:
+            self.collectionLayout.secondSectionBottomY = 390
+            self.collectionLayout.spacingKeyPoints = [0, 40, 74, 86]
+        case .write:
+            self.collectionLayout.secondSectionBottomY = 260
+            self.collectionLayout.spacingKeyPoints = [0, 8, 16, 20]
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/ConversationMessagesCell.swift:136]: These are the magic values for the conversation message layout.

File: iOS/Flows/Conversation/Cells/MessagesTimeMachineCollectionViewLayout.swift [hunk 5]
```diff
@@ -192,12 +193,6 @@ class MessagesTimeMachineCollectionViewLayout: TimeMachineCollectionViewLayout {
     private var initialZPosition: CGFloat = 0
     /// Items that that were visible before the animation started.
     private var indexPathsVisibleBeforeAnimation: Set<IndexPath> = []
-    
-    override func prepare(forAnimatedBoundsChange oldBounds: CGRect) {
-        super.prepare(forAnimatedBoundsChange: oldBounds)
-        
-        self.shouldScrollToEnd = true
-    }
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/MessagesTimeMachineCollectionViewLayout.swift:200]: We shouldn't always scroll to the latest message when changing layouts. Instead we only do it when entering write mode.

File: iOS/Flows/Conversation/Cells/TimeMachineCollectionViewLayout.swift [hunk 6]
```diff
@@ -53,21 +53,14 @@ class TimeMachineCollectionViewLayout: UICollectionViewLayout {
     // MARK: - Layout Configuration
 
     /// The height of the cells.
-    var itemHeight: CGFloat
-    = MessageContentView.bubbleHeight + MessageDetailView.height + Theme.ContentOffset.short.value {
-        didSet { self.invalidateLayout() }
-    }
+    var itemHeight: CGFloat = 100
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/TimeMachineCollectionViewLayout.swift:56]: Classes using this layout can specify what they want.

File: iOS/Flows/Conversation/Cells/TimeMachineCollectionViewLayout.swift [hunk 7]
```diff
@@ -53,21 +53,14 @@ class TimeMachineCollectionViewLayout: UICollectionViewLayout {
...
-        switch self.uiState {
-        case .read:
-            return [0, 18, 22, 26]
-        case .write:
-            return [0, 8, 16, 20]
-        }
-    }
+    var spacingKeyPoints: [CGFloat] = [0, 8, 16, 20]
+    var firstSectionTopY: CGFloat = 0
+    var secondSectionBottomY: CGFloat = 300
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/TimeMachineCollectionViewLayout.swift:62]: This lets you more precisely position where the message sections appear on screen.

File: iOS/Flows/Conversation/Cells/TimeMachineCollectionViewLayout.swift [hunk 8]
```diff
@@ -353,55 +345,27 @@ class TimeMachineCollectionViewLayout: UICollectionViewLayout {
...
+        var centerPoint = CGPoint(x: contentRect.midX, y: contentRect.top)
+
+        if section == 0 {
+            centerPoint.y += self.firstSectionTopY + self.itemHeight.half
+            centerPoint.y += self.itemHeight.half - self.itemHeight.half * scale
+            centerPoint.y += yOffset
+        } else {
+            centerPoint.y += self.secondSectionBottomY - self.itemHeight.half
+            centerPoint.y -= self.itemHeight.half - self.itemHeight.half * scale
+            centerPoint.y -= yOffset
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/TimeMachineCollectionViewLayout.swift:368]: Now that you specify the section positions, the logic is a lot simpler and more generic.

File: iOS/Flows/Conversation/Conversation List/ConversationListViewController.swift [hunk 9]
```diff
@@ -148,26 +148,14 @@ class ConversationListViewController: ViewController {
...
-                    settable.set(state: state)
-                }
-            }
-            
-            await self.dataSource.reconfigureAllItems()
-            
-            await UIView.awaitAnimation(with: .fast, animations: {
-                self.view.layoutNow()
-            })
-        }.add(to: self.taskPool)
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Conversation List/ConversationListViewController.swift:164]: All of this is replaced with a call to reconfigure the items.

File: iOS/Flows/Threads/ThreadCollectionView.swift [hunk 10]
```diff
@@ -42,59 +42,3 @@ extension ThreadCollectionView: MessageSendingCollectionViewType {
         return .zero
     }
 }
-
-class ThreadTimeMachineCollectionViewLayout: MessagesTimeMachineCollectionViewLayout {
```
Comment (MartinJibber) [inline iOS/Flows/Threads/ThreadCollectionView.swift:46]: No need to have a subclass for threads. We can configure the base class as needed.

**Review by Benjamindodgson — APPROVED** [review]

**Pull request merged**